### PR TITLE
feat: display latest post from a specific tag on the homepage

### LINKED ISSUE #9590: Implement latest post for homepage grid.
Implementation follow-up issue for #9404.

### PR DESCRIPTION
**Pull request by mamieorine**
Fixes #9590

Changes proposed in this pull request:
- Update the card component to display latest post from specific tags
- Update the homepage grid to display latest post from specific tags on each section
- Update the document on [/handbook/markup-cards-collection/](https://web.dev/handbook/markup-cards-collection/)

### CODE REVIEW

**Review by jimjiminyjimjim — COMMENTED** [review]
LGTM

**Review by matthiasrohmer — COMMENTED** [review]
I only quickly glanced at it, two questions: does that cover the case `featured == latest` and setting both a featured and a latest post?

**Comment by mamieorine** [comment]
> I only quickly glanced at it, two questions: does that cover the case `featured == latest` and setting both a featured and a latest post?

Do you mean the case that the author setting up to show the featured post and the latest post, in which both the featured and latest one refer to the same post? 

Actually, there is no frontmatter key for displaying featured posts yet. There are 3 ways of showing the post on the homepage grid.
- The author needs to define the title, description, eyebrow, etc, by themselves.
- Find by URL. The card will automatically pull post data when setting `url: /[my-post]/`.
- Display the latest post when set `isLatestPost: true`.

So I think the author can avoid displaying the duplicate post in each tag by the available card setup that I mentioned above.

Also, if they want to display a specific post as the featured post, they can set up the eyebrow like this code sample below:

```
---
cards:
  - url: /terra-dark-mode/
    eyebrow:
      icon: 'featured'
      text: 'Featured'
    cardLayout: 'vertical'
---
```

File: src/site/_filters/filter-by-tag.js
Comment (matthiasrohmer) [inline src/site/_filters/filter-by-tag.js:22]: We have `EleventyPage` as a type 🙂

File: src/site/_filters/filter-by-tag.js
Comment (matthiasrohmer) [inline src/site/_filters/filter-by-tag.js:23]: `string[]`.

File: src/site/_filters/filter-by-tag.js
Comment (matthiasrohmer) [inline src/site/_filters/filter-by-tag.js:45]: Two things: actually sorting after filtering would be quicker, because you can assume that you have less elements to sort after filtering. But I wouldn't expect the method to sort at all, that's not what the name says.

File: src/site/_includes/partials/cards-collection.njk
Comment (matthiasrohmer) [inline src/site/_includes/partials/cards-collection.njk:13]: This would need to go through the `i18` filter.

File: src/site/content/en/handbook/markup-cards-collection/index.md
Comment (matthiasrohmer) [inline src/site/content/en/handbook/markup-cards-collection/index.md:23]: What do you think of combining those two keys? So if you set `tags` it's simply the latest one?

File: src/site/content/en/index.md
Comment (matthiasrohmer) [inline src/site/content/en/index.md:94]: Was this a test? 🙂

File: src/site/_includes/partials/cards-collection.njk
Comment (matthiasrohmer) [inline src/site/_includes/partials/cards-collection.njk:10]: I think I would move most of this logic to a filter named `latestPostByTags`, which uses `this.ctx.collections[tag]` to access the already created, memoized and sorted collections.

**Review by matthiasrohmer — COMMENTED** [review]
Left some comments. The essential bit is: we already got collections per tag and shouldn't recreate them.

File: src/site/_filters/filter-by-tag.js
Comment (matthiasrohmer) [inline src/site/_filters/filter-by-tag.js:34]: These are wild three lines 😅  I think what you are looking for is `Array.prototype.map`, since you are creating an array from another array.

L33 is problematic in two ways: you are destructuring an array (`...posts`) just to add one element, while you could instead add to the existing array (`Array.prototype.push`). But the other problematic bit is `.pop()`, as it alters the original array, which we do not own in this context - so that's evil.

I think this can be rewritten as:

```
const posts = tags.map((tag) => {
  const tagPosts = this.ctx.collections[tag];
  return tagPosts[tagPosts.length - 1];
}
```

But you can actually improve this even more and get rid of the sorting, if you use an intermediate variable to store current newest post, like this:

```js
let newestPost = null;
for (const tag of tags) {
  const tagPosts = this.ctx.collections[tag];
  const candidatePost = tagPosts[tagPosts.length - 1];

  // Accounts for the first run of the loop
  if (!newestPost) { 
    newestPost = candidatePost; 
    return;
  }

  // Check if the current candidate post is newer then the newest
  // post that is currently stored
  if (new Date(newestPost.date) > new Date(candidatePost.date)) {
    newestPost = candidatePost;
  }
}
```

That's slightly more complicated to read, but does the job in one loop (while `map` and `sort` would be two). Basically a fun interview question...

File: src/site/content/en/index.md
Comment (matthiasrohmer) [reply src/site/content/en/index.md:92]: Why I am asking: `url` is hardcoded to `/top-cwv-2023/`.

File: src/site/_filters/filter-by-tag.js
Comment (mamieorine) [reply src/site/_filters/filter-by-tag.js]: I have refactored this logic. Thank you for your suggestion.

File: src/site/_filters/filter-by-tag.js
Comment (matthiasrohmer) [inline src/site/_filters/filter-by-tag.js:38]: Never trust foreign code without testing 😄 my suggestion had an error that renders all "Latest" tiles empty. This needs to be `continue;`.

File: src/site/_filters/filter-by-tag.js
Comment (matthiasrohmer) [inline src/site/_filters/filter-by-tag.js:1]: The main export of this file now is `latestPostByTags`, the file should hence be named `latest-post-by-tags.js`.

**Review by matthiasrohmer — CHANGES_REQUESTED** [review]
Almost there. Please reset all changes to index.md, except the CSS and UI category, as in the other categories there is now some overlap between Featured and Latest that I the content team is better suited to sort out. CSS/UI on the other hand seems to be safe and is good to demonstrate how this works. Thanks!

File: src/site/_filters/filter-by-tag.js
Comment (mamieorine) [reply src/site/_filters/filter-by-tag.js]: my bad, I did not notice this bug.

File: src/site/_filters/filter-by-tag.js
Comment (mamieorine) [reply src/site/_filters/filter-by-tag.js]: I have renamed this file from `filter-by-tag.js` to `latest-post-by-tags.js`

**Review by matthiasrohmer — APPROVED** [review]
This works now. Forwarding to @rachelandrew for visibility and approval of making the CSS & UI theme use the latest post.

**Review by rachelandrew — APPROVED** [review]
LGTM - thanks! Once this is landed I'll look at updating the others.

**Pull request merged**
